### ISSUE TITLE
thread 'coordinator' panicked at 'assertion failed: !self.expressions.iter().any(|e| e.contains_temporal())', src/expr/src/linear.rs:461:9

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#18045]
### What version of Materialize are you using?

v0.46.0-dev (7394158)

### How did you install Materialize?

Built from source

### What is the issue?

```
materialize=> explain select from(select c0,
                    "c1" c10 from(select c0, "egress_ip" c1 from mz_egress_ips,
                                  lateral(select "egress_ip" c0))) where
        case when(cast(null as bpchar) ~"c10") then cast(null as uuid) end
        >= uuid_generate_v5(CAST(cast(cast(null as uuid) as uuid) as uuid),
                            CAST("c0" as text)) or
    cast(nullif(cast(nullif(mz_now(),
                            (select "time" from mz_internal.mz_active_peeks))
                         as mz_timestamp),
                case when NULL then 0 end) as mz_timestamp) > case when
        mz_now() >=
        0 then(select "time" from mz_internal.mz_worker_compute_frontiers) end;
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
```

### Relevant log output

```source
thread 'coordinator' panicked at 'assertion failed: !self.expressions.iter().any(|e| e.contains_temporal())', src/expr/src/linear.rs:461:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fdbc4329cb781c7768ffa6d76c8fa2d032d3fe20/library/std/src/panicking.rs:579:5
   1: core::panicking::panic_fmt
             at /rustc/fdbc4329cb781c7768ffa6d76c8fa2d032d3fe20/library/core/src/panicking.rs:64:14
   2: core::panicking::panic
             at /rustc/fdbc4329cb781c7768ffa6d76c8fa2d032d3fe20/library/core/src/panicking.rs:114:5
   3: mz_expr::linear::MapFilterProject::extract_temporal
             at ./src/expr/src/linear.rs:461:9
   4: mz_compute_client::plan::join::delta_join::DeltaJoinPlan::create_from
             at ./src/compute-client/src/plan/join/delta_join.rs:249:28
   5: mz_compute_client::plan::Plan<T>::from_mir_stack_safe
             at ./src/compute-client/src/plan/mod.rs:1234:46
   6: mz_compute_client::plan::Plan<T>::from_mir_inner::{{closure}}
             at ./src/compute-client/src/plan/mod.rs:952:38
   7: stacker::maybe_grow
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
   8: mz_ore::stack::maybe_grow
             at ./src/ore/src/stack.rs:100:5
   9: mz_compute_client::plan::Plan<T>::from_mir_inner
             at ./src/compute-client/src/plan/mod.rs:952:9
  10: mz_compute_client::plan::Plan<T>::from_mir_stack_safe
             at ./src/compute-client/src/plan/mod.rs:1448:21
  11: mz_compute_client::plan::Plan<T>::from_mir_inner::{{closure}}
             at ./src/compute-client/src/plan/mod.rs:952:38
  12: stacker::maybe_grow
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  13: mz_ore::stack::maybe_grow
             at ./src/ore/src/stack.rs:100:5
  14: mz_compute_client::plan::Plan<T>::from_mir_inner
             at ./src/compute-client/src/plan/mod.rs:952:9
  15: mz_compute_client::plan::Plan<T>::from_mir_stack_safe
             at ./src/compute-client/src/plan/mod.rs:1175:40
  16: mz_compute_client::plan::Plan<T>::from_mir_inner::{{closure}}
             at ./src/compute-client/src/plan/mod.rs:952:38
  17: stacker::maybe_grow
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  18: mz_ore::stack::maybe_grow
             at ./src/ore/src/stack.rs:100:5
  19: mz_compute_client::plan::Plan<T>::from_mir_inner
             at ./src/compute-client/src/plan/mod.rs:952:9
  20: mz_compute_client::plan::Plan<T>::from_mir_stack_safe
             at ./src/compute-client/src/plan/mod.rs:1064:38
  21: mz_compute_client::plan::Plan<T>::from_mir_inner::{{closure}}
             at ./src/compute-client/src/plan/mod.rs:952:38
  22: stacker::maybe_grow
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  23: mz_ore::stack::maybe_grow
             at ./src/ore/src/stack.rs:100:5
  24: mz_compute_client::plan::Plan<T>::from_mir_inner
             at ./src/compute-client/src/plan/mod.rs:952:9
  25: mz_compute_client::plan::Plan<T>::from_mir_stack_safe
             at ./src/compute-client/src/plan/mod.rs:1064:38
  26: mz_compute_client::plan::Plan<T>::from_mir_inner::{{closure}}
             at ./src/compute-client/src/plan/mod.rs:952:38
  27: stacker::maybe_grow
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  28: mz_ore::stack::maybe_grow
             at ./src/ore/src/stack.rs:100:5
  29: mz_compute_client::plan::Plan<T>::from_mir_inner
             at ./src/compute-client/src/plan/mod.rs:952:9
  30: mz_compute_client::plan::Plan<T>::from_mir_stack_safe
             at ./src/compute-client/src/plan/mod.rs:1064:38
  31: mz_compute_client::plan::Plan<T>::from_mir_inner::{{closure}}
             at ./src/compute-client/src/plan/mod.rs:952:38
  32: stacker::maybe_grow
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.15/src/lib.rs:55:9
  33: mz_ore::stack::maybe_grow
             at ./src/ore/src/stack.rs:100:5
  34: mz_compute_client::plan::Plan<T>::from_mir_inner
             at ./src/compute-client/src/plan/mod.rs:952:9
  35: mz_compute_client::plan::Plan<T>::from_mir
             at ./src/compute-client/src/plan/mod.rs:935:9
  36: mz_compute_client::plan::Plan<T>::finalize_dataflow
             at ./src/compute-client/src/plan/mod.rs:1612:32
  37: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_explain_plan::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/sequencer.rs:2926:17
  38: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_explain_plan::catch_unwind
             at ./src/adapter/src/coord/sequencer.rs:2854:17
  39: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_explain_plan::{{closure}}
             at ./src/adapter/src/coord/sequencer.rs:2925:33
  40: tracing_core::dispatcher::with_default
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-core-0.1.30/src/dispatcher.rs:257:5
  41: mz_adapter::explain::optimizer_trace::OptimizerTrace::collect_trace
             at ./src/adapter/src/explain/optimizer_trace.rs:77:9
  42: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_explain_plan
             at ./src/adapter/src/coord/sequencer.rs:2865:31
  43: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_explain
             at ./src/adapter/src/coord/sequencer.rs:2812:26
  44: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_plan::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/sequencer.rs:337:17
  45: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator>::sequence_plan::{{closure}}
             at ./src/adapter/src/coord/sequencer.rs:118:5
  46: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute_inner::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:504:78
  47: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute_inner::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:307:5
  48: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:304:61
  49: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_execute::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:263:5
  50: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
  51: mz_adapter::coord::command_handler::<impl mz_adapter::coord::Coordinator>::handle_command::{{closure}}
             at ./src/adapter/src/coord/command_handler.rs:68:21
  52: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::message_command::{{closure}}::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:249:33
  53: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::message_command::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:246:5
  54: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator>::handle_message::{{closure}}
             at ./src/adapter/src/coord/message_handler.rs:44:63
  55: mz_adapter::coord::Coordinator::serve::{{closure}}
             at ./src/adapter/src/coord.rs:1298:37
  56: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/park.rs:283:63
  57: tokio::runtime::coop::with_budget
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/coop.rs:102:5
  58: tokio::runtime::coop::budget
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/coop.rs:68:5
  59: tokio::runtime::park::CachedParkThread::block_on
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/park.rs:283:31
  60: tokio::runtime::context::BlockingRegionGuard::block_on
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/context.rs:315:13
  61: tokio::runtime::handle::Handle::block_on
             at /home/deen/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.2/src/runtime/handle.rs:264:9
  62: mz_adapter::coord::serve::{{closure}}::{{closure}}
             at ./src/adapter/src/coord.rs:1477:17
```

**Comment by antiguru** [comment]
Cc @aalexandrov @ggevay

**Comment by aalexandrov** [comment]
Taking a look.

**Comment by aalexandrov** [comment]
I managed to simplify this to the following statement:

```sql
select 
  *
from
  mz_egress_ips
where
  nullif(nullif(mz_now(), (select "time" from mz_internal.mz_active_peeks limit 1)), NULL)
  > 
  (select "time" from mz_internal.mz_worker_compute_frontiers limit 1);
```

The `explain` of this fails with the above error, but regular `select` works.

**Comment by aalexandrov** [comment]
TL;DR: we have a situation where an expression that starts in the `WHERE` clause actually ends up in the `Map` that is not  followed by a project so `MapFilterProject` handling during lowering cannot move it back to be part of a `Filter`. Consequently, the

```rust
assert!(!self.expressions.iter().any(|e| e.contains_temporal()));
```

assertion from the code snippet referenced below fails. @frankmcsherry any suggestions how to go about fixing that? I believe that the issue won't manifest if the offending optimized plan

```text
Explained Query:
  Return
    Project (#0)
      Filter (#1 > #2)
        CrossJoin type=differential
          ArrangeBy keys=[[]]
            Project (#0, #2)
              Map (case when (#1 = mz_now()) then null else mz_now() end)
                CrossJoin type=differential
                  ArrangeBy keys=[[]]
                    Get mz_catalog.mz_egress_ips
                  ArrangeBy keys=[[]]
                    Union
                      Get l0
                      Map (null)
                        Union
                          Negate
                            Project ()
                              Get l0
                          Constant
                            - ()
          ArrangeBy keys=[[]]
            TopK limit=1 monotonic=false
              Project (#2)
                Get mz_internal.mz_worker_compute_frontiers
  With
    cte l0 =
      TopK limit=1 monotonic=false
        Project (#3)
          Get mz_internal.mz_active_peeks
```

is rewritten in a way such that the 

```text
Map (case when (#1 = mz_now()) then null else mz_now() end)
```

is inlined in the `#1` reference in

```text
Filter (#1 > #2)
```

Should we always try to lift scalar expressions from `maps` to their (assumed exclusive use) in `filter` predicate references?

**Comment by aalexandrov** [comment]
Here is what happens if we try to materialize a slightly adjusted version of the query (I think this is the real issue):

```sql
mz> create table t1(x int);
CREATE TABLE
Time: 0.192s

mz> create materialized view mv as
 select
   *
 from
   t1
 where
   nullif(nullif(mz_now(), (select x from t1 limit 1)), NULL)
   >
   (select x * 2 from t1 limit 1);
cannot materialize call to mz_now
```

**Comment by antiguru** [comment]
We put this work on hold until we have more evidence that it's affecting customers. We'll try to fix it once we have some spare capacity.
